internal/v4/api.go

```diff
@@ -787,13 +788,46 @@ func (h *ReqHandler) putMetaPerm(id *router.ResolvedURL, path string, val *json.
 			break
 		}
 	}
+
 	updater.UpdateField("acls.read", perms.Read)
 	updater.UpdateField("public", isPublic)
 	updater.UpdateField("acls.write", perms.Write)
+
+	// This is currently wrong as the updater will fire later and might fail.
+	// TODO: Associate the audit entry with the FieldUpdater
+	h.addAuditSetPerms(id, perms.Read, perms.Write)
+
 	updater.UpdateSearch()
 	return nil
 }
 
+var testAddAuditCallback func(e audit.Entry)
+
+// addAuditSetPerms adds an audit entry recording that the permissions of the given
+// entity have been set to the given ACLs.
+func (h *ReqHandler) addAuditSetPerms(id *router.ResolvedURL, read, write []string) {
+	if h.auth.Username == "" && !h.auth.Admin {
+		panic("No auth set in ReqHandler")
+	}
+	e := audit.Entry{
+		Op:     audit.OpSetPerm,
+		Entity: &id.URL,
+		ACL: &audit.ACL{
+			Read:  read,
+			Write: write,
+		},
+		User: h.auth.Username,
+	}
+	if h.auth.Admin && e.User == "" {
+		e.User = "admin"
+	}
+	h.Store.AddAudit(e)
+
+	if testAddAuditCallback != nil {
```

Comment:
Me too, but I *think* this is better than making the v4 API depend on details
of the audit log output. Alternative suggestions welcome.